packages/gravity/kube-testing/src/spec/replication.ts

```diff
@@ -0,0 +1,412 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { randomBytes } from 'node:crypto';
+
+import { sleep, scheduleTaskInterval } from '@dxos/async';
+import { Context } from '@dxos/context';
+import { FeedFactory, FeedStore, type FeedWrapper } from '@dxos/feed-store';
+import { Keyring, TestKeyPair, generateKeyPair } from '@dxos/keyring';
+import { PublicKey } from '@dxos/keys';
+import { log } from '@dxos/log';
+import { TestBuilder as NetworkManagerTestBuilder } from '@dxos/network-manager/testing';
+import { createStorage, StorageType } from '@dxos/random-access-storage';
+import { ReplicatorExtension } from '@dxos/teleport-extension-replicator';
+import { range } from '@dxos/util';
+
+import { SerializedLogEntry, getReader } from '../analysys';
+import { PlanResults, TestParams, TestPlan, AgentEnv } from '../plan';
+import { TestBuilder as SignalTestBuilder } from '../test-builder';
+import { forEachSwarmAndAgent, joinSwarm, leaveSwarm } from './util';
+
+const REPLICATOR_EXTENSION_NAME = 'replicator';
+
+type FeedConfig = {
+  feedKey: string;
+  writable: boolean;
+};
+
+type FeedEntry = {
+  agentIdx: number;
+  swarmIdx: number;
+  feedLength: number;
+  byteLength: number;
+  writable: boolean;
+  feedKey: string;
+};
+
+type FeedReplicaStats = {
+  feedLength: number;
+  byteLength: number;
+  replicationSpeed?: string;
+};
+
+type FeedStats = {
+  feedKey: string;
+  feedLength: number;
+  byteLength: number;
+  replicas: FeedReplicaStats[];
+};
+
+export type ReplicationTestSpec = {
+  agents: number;
+  swarmsPerAgent: number;
```

Comment:
not sure if having multiple swarms is useful in this tests, how is it different from just adding more feeds?